lib/elixir/lib/calendar.ex

```diff
@@ -413,13 +513,54 @@ defmodule Date do
   """
   @spec compare(Calendar.date, Calendar.date) :: :lt | :eq | :gt
   def compare(date1, date2) do
-    case {to_erl(date1), to_erl(date2)} do
-      {first, second} when first > second -> :gt
-      {first, second} when first < second -> :lt
-      _ -> :eq
+    if date1.calendar.day_rollover_relative_to_midnight_utc == date2.calendar.day_rollover_relative_to_midnight_utc do
```

Comment:
Move the day rollover check to a separate function as there is no reason to run the check if they have the same calendar.